File: scripts/recrop-layering-ref.ts
/**
 * Re-crop a layering ref to the hem ↔ shoe region only.
 *
 * Motivation (2026-05-24): Cargo trouser test scored 2/6 because the
 * substitute layering ref (Rovic Cargo) differs from the actual cargo
 * on wash + silhouette + cut. Seedream pulled those conflicting signals
 * from the ref instead of using it only for hem-vs-shoe geometry.
 *
 * Crop to bottom 40% (below-knee → floor) keeps the ONLY signal we
 * actually want — the geometric relationship between trouser hem and
 * shoe — and drops the noise. Floor stays visible at the bottom of
 * the crop so Seedream reads it as a hem detail, not a cropped pant.
 *
 * Overwrites the same GCS path (gs://gstar-ai-studio-assets/layering-refs/
 * {wardrobeId}.jpg) so no other code needs to change — test-m02-tier2-arch.ts
 * picks up the new image automatically on the next run.
 *
 * Usage:
 *   npx tsx scripts/recrop-layering-ref.ts <wardrobeId> [--region=below-knee]
 *   npx tsx scripts/recrop-layering-ref.ts --all-substitutes [--region=below-knee]
 *
 * Regions:
 *   below-knee (default) — bottom 40% of source. Knee → hem → shoe → floor.
 *   tight                 — bottom 28%. Mid-calf → shoe → floor. More aggressive.
 *
 * Backed up to gs://gstar-ai-studio-assets/layering-refs/full/{wardrobeId}.jpg
 * before overwrite so restore is one gsutil mv away.
 */
import * as fs from 'fs';
import * as path from 'path';

const projectRoot = path.resolve(__dirname, '..');
const envPath = path.join(projectRoot, '.env.local');
if (fs.existsSync(envPath)) {
  for (const line of fs.readFileSync(envPath, 'utf-8').split('\n')) {
    const m = line.match(/^([A-Z_][A-Z0-9_]*)=(.*)$/);
    if (m && !process.env[m[1]]) process.env[m[1]] = m[2].replace(/^['"]|['"]$/g, '');
  }
}
process.env.GOOGLE_APPLICATION_CREDENTIALS = path.join(projectRoot, 'sa_key.json');

import { Firestore } from '@google-cloud/firestore';
import { Storage } from '@google-cloud/storage';
import sharp from 'sharp';

const BUCKET = 'gstar-ai-studio-assets';
const ARCHIVE_PREFIX = 'layering-refs/full';
const LIVE_PREFIX = 'layering-refs';

const args = process.argv.slice(2);
const allSubstitutes = args.includes('--all-substitutes');
const regionArg = args.find(a => a.startsWith('--region=')) || '--region=below-knee';
const region = regionArg.split('=')[1] as 'below-knee' | 'tight';
const wardrobeId = !allSubstitutes ? args[0] : null;

if (!wardrobeId && !allSubstitutes) {
  console.error('Usage: npx tsx scripts/recrop-layering-ref.ts <wardrobeId> [--region=below-knee|tight]');
  console.error('   or: npx tsx scripts/recrop-layering-ref.ts --all-substitutes [--region=below-knee|tight]');
  process.exit(1);
}

const REGION_BOTTOM = { 'below-knee': 0.40, 'tight': 0.28 }[region];
if (!REGION_BOTTOM) {
  console.error(`Unknown region: ${region}. Use 'below-knee' or 'tight'.`);
  process.exit(1);
}

async function process(db: Firestore, storage: Storage, wid: string): Promise<{ok: boolean; note?: string}> {
  const docRef = db.collection('wardrobe').doc(wid);
  const snap = await docRef.get();
  if (!snap.exists) return { ok: false, note: 'wardrobe doc not found' };
  const data = snap.data() as any;
  const livePath = `${LIVE_PREFIX}/${wid}.jpg`;
  const archivePath = `${ARCHIVE_PREFIX}/${wid}.jpg`;

  const bucket = storage.bucket(BUCKET);
  const liveFile = bucket.file(livePath);
  const [exists] = await liveFile.exists();
  if (!exists) return { ok: false, note: 'GCS file not found' };

  // 1. Archive the current full image (idempotent — skips if already archived)
  const archiveFile = bucket.file(archivePath);
  const [archived] = await archiveFile.exists();
  if (!archived) {
    await liveFile.copy(archiveFile);
    console.log(`  archived → gs://${BUCKET}/${archivePath}`);
  } else {
    console.log(`  already archived (skip archive step)`);
  }

  // 2. Download, crop, re-upload
  const [buf] = await liveFile.download();
  const meta = await sharp(buf).metadata();
  const w = meta.width!, h = meta.height!;
  const cropH = Math.round(h * REGION_BOTTOM);
  const topY = h - cropH;
  const cropped = await sharp(buf)
    .extract({ left: 0, top: topY, width: w, height: cropH })
    .jpeg({ quality: 92 })
    .toBuffer();
  await liveFile.save(cropped, {
    metadata: { contentType: 'image/jpeg', cacheControl: 'public, max-age=3600' },
    resumable: false,
  });

  // 3. Update Firestore with a cache-buster + flag
  const newUrl = `https://storage.googleapis.com/${BUCKET}/${livePath}?v=${Date.now()}`;
  const existingSource = data.layeringRefSource || {};
  await docRef.update({
    layeringRefBackUrl: newUrl,
    layeringRefSource: {
      ...existingSource,
      cropped: true,
      cropRegion: region,
      croppedAt: new Date().toISOString(),
    },
    updatedAt: new Date(),
  });

  console.log(`  ✓ recropped ${region} (${w}×${h} → ${w}×${cropH}, ${(cropped.length/1024).toFixed(0)} KB)`);
  return { ok: true };
}

async function main() {
  const saKey = JSON.parse(fs.readFileSync(path.join(projectRoot, 'sa_key.json'), 'utf-8'));
  const db = new Firestore({
    projectId: saKey.project_id || 'gstar-ai-studio',
    credentials: { client_email: saKey.client_email, private_key: saKey.private_key },
  });
  const storage = new Storage({
    projectId: saKey.project_id || 'gstar-ai-studio',
    credentials: { client_email: saKey.client_email, private_key: saKey.private_key },
  });

  console.log(`Mode: ${allSubstitutes ? 'ALL substitutes' : `single (${wardrobeId})`} | region: ${region} (bottom ${(REGION_BOTTOM*100).toFixed(0)}%)`);

  if (allSubstitutes) {
    // Iterate wardrobe items where layeringRefSource.substitute === true
    const snap = await db.collection('wardrobe').where('category', '==', 'bottom').get();
    let processed = 0, skipped = 0, failed = 0;
    for (const doc of snap.docs) {
      const data = doc.data() as any;
      const src = data.layeringRefSource || {};
      if (!src.substitute) { skipped++; continue; }
      console.log(`\n${doc.id}  ${data.name}  (substitute for ${src.substituteFor})`);
      try {
        const r = await process(db, storage, doc.id);
        if (r.ok) processed++; else { failed++; console.log(`  ⚠️  ${r.note}`); }
      } catch (e) {
        failed++;
        console.log(`  ❌ ${(e as Error).message}`);
      }
    }
    console.log(`\nSummary: ${processed} cropped, ${skipped} skipped (not substitutes), ${failed} failed.`);
  } else {
    console.log(`\n${wardrobeId}`);
    const r = await process(db, storage, wardrobeId!);
    if (!r.ok) {
      console.error(`Failed: ${r.note}`);
      process.exit(1);
    }
  }
  console.log(`\n💡 Re-run test-m02-tier2-arch.ts against the same job(s) to see if the cropped ref improves stability.`);
  console.log(`💡 To restore originals: gsutil cp gs://${BUCKET}/${ARCHIVE_PREFIX}/<wardrobeId>.jpg gs://${BUCKET}/${LIVE_PREFIX}/<wardrobeId>.jpg`);
  process.exit(0);
}

main().catch(e => { console.error(e); process.exit(1); });

File: scripts/recrop_layering_ref.py
#!/usr/bin/env python3
"""
Re-crop a layering ref to the hem ↔ shoe region only.

Same goal as recrop-layering-ref.ts (which was tripping Node 25 + tsx)
but written in Python to bypass the Node toolchain entirely. Uses the
same sa_key.json + GCS bucket as the rest of the scripts/.

The point: when a wardrobe item uses a substitute layering ref (a
different SKU from the actual garment), Seedream pulls wash/silhouette/
pocket signals from the full-image ref and renders the wrong garment.
Cropping to the bottom 40% (below-knee → floor) keeps only the hem ↔
shoe geometry — the only signal we actually want.

Overwrites gs://gstar-ai-studio-assets/layering-refs/{wid}.jpg so
test-m02-tier2-arch.ts picks up the new image on the next run. The
original full image is archived to layering-refs/full/{wid}.jpg
before overwrite (idempotent).

Usage:
  python3 scripts/recrop_layering_ref.py <wardrobeId> [--region=below-knee|tight]
  python3 scripts/recrop_layering_ref.py --all-substitutes [--region=below-knee|tight]

Regions:
  below-knee (default)  bottom 40% of source
  tight                  bottom 28%

Restore originals: gsutil cp gs://.../layering-refs/full/<wid>.jpg .../layering-refs/<wid>.jpg
"""
import argparse
import io
import json
import os
import sys
import time
from pathlib import Path

# Resolve project root + ensure sa_key.json is on the credentials path
PROJECT_ROOT = Path(__file__).resolve().parent.parent
SA_KEY_PATH = PROJECT_ROOT / "sa_key.json"
if not SA_KEY_PATH.exists():
    print(f"sa_key.json not found at {SA_KEY_PATH}", file=sys.stderr)
    sys.exit(1)
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = str(SA_KEY_PATH)

# Lazy-import so error messages are clearer if a dep is missing
try:
    from PIL import Image
except ImportError:
    print("Pillow not installed. Run: pip3 install --user Pillow google-cloud-firestore google-cloud-storage", file=sys.stderr)
    sys.exit(1)
try:
    from google.cloud import firestore, storage
except ImportError:
    print("google-cloud-* not installed. Run: pip3 install --user google-cloud-firestore google-cloud-storage", file=sys.stderr)
    sys.exit(1)

BUCKET = "gstar-ai-studio-assets"
LIVE_PREFIX = "layering-refs"
ARCHIVE_PREFIX = "layering-refs/full"

REGION_BOTTOM = {
    "below-knee": 0.40,
    "tight": 0.28,
}

def process_one(db, bucket, wardrobe_id: str, region: str) -> tuple[bool, str | None]:
    region_frac = REGION_BOTTOM[region]
    doc_ref = db.collection("wardrobe").document(wardrobe_id)
    snap = doc_ref.get()
    if not snap.exists:
        return False, "wardrobe doc not found"
    data = snap.to_dict()

    live_path = f"{LIVE_PREFIX}/{wardrobe_id}.jpg"
    archive_path = f"{ARCHIVE_PREFIX}/{wardrobe_id}.jpg"
    live_blob = bucket.blob(live_path)
    if not live_blob.exists():
        return False, "GCS file not found"

    # 1. Archive the current full image once (idempotent)
    archive_blob = bucket.blob(archive_path)
    if not archive_blob.exists():
        bucket.copy_blob(live_blob, bucket, archive_path)
        print(f"  archived → gs://{BUCKET}/{archive_path}")
    else:
        print(f"  already archived (skip)")

    # 2. Download, crop, re-upload
    src_bytes = live_blob.download_as_bytes()
    img = Image.open(io.BytesIO(src_bytes)).convert("RGB")
    w, h = img.size
    crop_h = int(h * region_frac)
    cropped = img.crop((0, h - crop_h, w, h))
    out_buf = io.BytesIO()
    cropped.save(out_buf, format="JPEG", quality=92)
    out_bytes = out_buf.getvalue()
    live_blob.cache_control = "public, max-age=3600"
    live_blob.upload_from_string(out_bytes, content_type="image/jpeg")

    # 3. Update Firestore — cache-buster + flag
    new_url = f"https://storage.googleapis.com/{BUCKET}/{live_path}?v={int(time.time() * 1000)}"
    existing_source = data.get("layeringRefSource") or {}
    doc_ref.update({
        "layeringRefBackUrl": new_url,
        "layeringRefSource": {
            **existing_source,
            "cropped": True,
            "cropRegion": region,
            "croppedAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        },
        "updatedAt": firestore.SERVER_TIMESTAMP,
    })
    print(f"  ✓ recropped {region} ({w}×{h} → {w}×{crop_h}, {len(out_bytes)//1024} KB)")
    return True, None


def main():
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("wardrobe_id", nargs="?", help="Wardrobe item ID to recrop")
    parser.add_argument("--all-substitutes", action="store_true", help="Iterate all bottoms with substitute layering refs")
    parser.add_argument("--region", choices=list(REGION_BOTTOM.keys()), default="below-knee")
    args = parser.parse_args()

    if not args.wardrobe_id and not args.all_substitutes:
        parser.error("provide a <wardrobeId> or --all-substitutes")

    with open(SA_KEY_PATH) as f:
        sa_key = json.load(f)
    project_id = sa_key.get("project_id", "gstar-ai-studio")
    db = firestore.Client(project=project_id)
    storage_client = storage.Client(project=project_id)
    bucket = storage_client.bucket(BUCKET)

    print(f"Mode: {'ALL substitutes' if args.all_substitutes else f'single ({args.wardrobe_id})'} | region: {args.region} (bottom {int(REGION_BOTTOM[args.region]*100)}%)")

    if args.all_substitutes:
        col = db.collection("wardrobe").where("category", "==", "bottom").stream()
        processed = skipped = failed = 0
        for doc in col:
            data = doc.to_dict()
            src = data.get("layeringRefSource") or {}
            if not src.get("substitute"):
                skipped += 1
                continue
            print(f"\n{doc.id}  {data.get('name')}  (substitute for {src.get('substituteFor')})")
            try:
                ok, note = process_one(db, bucket, doc.id, args.region)
                if ok: processed += 1
                else: failed += 1; print(f"  ⚠️  {note}")
            except Exception as e:
                failed += 1
                print(f"  ❌ {e}")
        print(f"\nSummary: {processed} cropped, {skipped} skipped (not substitutes), {failed} failed.")
    else:
        print(f"\n{args.wardrobe_id}")
        ok, note = process_one(db, bucket, args.wardrobe_id, args.region)
        if not ok:
            print(f"Failed: {note}", file=sys.stderr)
            sys.exit(1)

    print(f"\n💡 Re-run test-m02-tier2-arch.ts to see if the cropped ref improves stability.")
    print(f"💡 Restore: gsutil cp gs://{BUCKET}/{ARCHIVE_PREFIX}/<wid>.jpg gs://{BUCKET}/{LIVE_PREFIX}/<wid>.jpg")


if __name__ == "__main__":
    main()
